### PR TITLE
[AI] Fix Escape to exit edit mode and clear multi-select in tables

### PR DESCRIPTION
**Pull request by nikhilweee**
## Description

Clicking a cell in the transactions, budget, payees, or tags table jumps directly into edit mode, but pressing **Escape** previously did nothing useful: in plain text cells it would revert the typed value but the input stayed focused; in autocomplete or date-picker cells it didn't even do that. Multi-selecting rows via checkboxes also had no keyboard escape — you had to click the checkboxes again or use the toolbar to clear the selection.

This PR makes Escape a proper "back out" key:

- Inside an editable cell: Escape exits edit mode and reverts the typed value.
- With rows checkbox-selected (and no cell being edited): Escape clears the selection in a single press.
- When both states overlap (a cell is being edited *and* rows are checked): the first Escape exits edit, the second clears the selection.
- Popovers (autocomplete dropdown, date picker) keep their existing Escape-to-close behavior; their handlers stop propagation, so the new Escape logic only fires once the popover is closed.

## Related issue(s)

None.

## Testing

Manual:

1. Open the demo budget (View demo).
2. **Budget grid**: click any budgeted cell, type a new value, press Escape — cell exits edit mode and reverts to its previous value (the total budgeted at the top should not change).
3. **Transactions table** (open any account):
   - Click a notes/payee/category cell, type, press Escape — value reverts and cell exits.
   - Open a category cell with the autocomplete dropdown, press Escape — dropdown closes; press Escape again — cell exits.
   - Check 3 rows via checkboxes, press Escape — all unchecked, toolbar disappears.
   - Check rows, then click another cell to edit; press Escape — exit edit (rows still checked); press Escape again — rows unchecked.
4. **Payees** and **Tags** pages: same behavior where applicable.

Automated: new `packages/desktop-client/e2e/keyboard-navigation.test.ts` (6 tests) covers the budget grid (test budget + demo budget), transactions table (single ESC, multi-select clear, layered peel-off), and the cell-with-existing-value scenario that surfaces the value-revert path.

## Checklist

- [x] Release notes added (see link above)
- [x] No obvious regressions in affected areas
- [x] Self-review has been performed - I understand what each change in the code does and why it is needed

<hr />

### Bundle Stats

Bundle | Files count | Total bundle size | % Changed
------ | ----------- | ----------------- | ---------
desktop-client | 42 | 13.88 MB → 13.8 MB (-82.73 kB) | -0.58%
loot-core | 1 | 4.65 MB | 0%
api | 2 | 3.87 MB | 0%
cli | 1 | 8.02 MB | 0%
crdt | 1 | 11.16 kB | 0%

<details>
<summary>View detailed bundle stats</summary>

#### desktop-client

**Total**
Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
42 | 13.88 MB → 13.8 MB (-82.73 kB) | -0.58%

<details>
<summary>Changeset</summary>

File | Δ | Size
---- | - | ----
`locale/en-GB.json` | 📈 +2.59 kB (+23.31%) | 11.1 kB → 13.69 kB
`src/components/table.tsx` | 📈 +2.11 kB (+5.81%) | 36.23 kB → 38.33 kB
`src/components/select/DateSelect.tsx` | 📈 +267 B (+1.67%) | 15.64 kB → 15.9 kB
`src/components/budget/BudgetTable.tsx` | 📈 +79 B (+1.26%) | 6.11 kB → 6.18 kB
`src/components/autocomplete/TagAutocomplete.tsx` | 📈 +73 B (+1.13%) | 6.32 kB → 6.39 kB
`src/components/transactions/TransactionsTable.tsx` | 📈 +295 B (+0.31%) | 93.56 kB → 93.85 kB
</details>

<details>
<summary>View detailed bundle breakdown</summary>
<div>

**Added**
No assets were added

**Removed**
No assets were removed

**Bigger**
Asset | File Size | % Changed
----- | --------- | ---------
static/js/en-GB.js | 11.1 kB → 13.69 kB (+2.59 kB) | +23.31%
static/js/Value.js | 1.79 MB → 1.79 MB (+2.11 kB) | +0.11%
static/js/DateSelect.js | 2.37 MB → 2.37 MB (+267 B) | +0.01%
static/js/wide.js | 274.53 kB → 274.6 kB (+79 B) | +0.03%

**Smaller**
Asset | File Size | % Changed
----- | --------- | ---------
static/js/TransactionList.js | 166.98 kB → 79.22 kB (-87.76 kB) | -52.56%

**Unchanged**
Asset | File Size | % Changed
----- | --------- | ---------
static/js/index.js | 1.7 MB | 0%
static/js/BackgroundImage.js | 121.09 kB | 0%
static/js/FormulaEditor.js | 767.24 kB | 0%
static/js/ManageRules.js | 71.66 kB | 0%
static/js/NotesTagFormatter.js | 25.94 kB | 0%
static/js/PayeeRuleCountLabel.js | 12.81 kB | 0%
static/js/ReportRouter.js | 1.53 MB | 0%
static/js/ScheduleEditForm.js | 75.41 kB | 0%
static/js/SchedulesTable.js | 171.25 kB | 0%
static/js/TourHost.js | 169.22 kB | 0%
static/js/TourProvider.js | 1.54 kB | 0%
static/js/TransactionEdit.js | 219.59 kB | 0%
static/js/bootstrapHyperFormula.js | 302 B | 0%
static/js/ca.js | 170.64 kB | 0%
static/js/client.js | 452.05 kB | 0%
static/js/columns.js | 673.28 kB | 0%
static/js/de.js | 186.5 kB | 0%
static/js/en.js | 258.91 kB | 0%
static/js/es.js | 182.1 kB | 0%
static/js/fr.js | 178.66 kB | 0%
static/js/indexeddb-main-thread-worker-e59fee74.js | 13.46 kB | 0%
static/js/it.js | 151.91 kB | 0%
static/js/months.js | 590.14 kB | 0%
static/js/narrow.js | 290.52 kB | 0%
static/js/nb-NO.js | 135.46 kB | 0%
static/js/nl.js | 151.78 kB | 0%
static/js/pl.js | 136.51 kB | 0%
static/js/pt-BR.js | 178.74 kB | 0%
static/js/ru.js | 212.44 kB | 0%
static/js/theme.js | 31.68 kB | 0%
static/js/uk.js | 211.2 kB | 0%
static/js/useFormatList.js | 10.4 kB | 0%
static/js/useLocalPref.js | 97.2 kB | 0%
static/js/useReducedMotion.js | 594 B | 0%
static/js/workbox-window.prod.es5.js | 7.21 kB | 0%
static/js/zh-Hans.js | 106.8 kB | 0%
static/js/zh-Hant.js | 130.92 kB | 0%
</div>
</details>

---

#### loot-core

**Total**
Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
1 | 4.65 MB | 0%

<details>
<summary>View detailed bundle breakdown</summary>
<div>

**Added**
No assets were added

**Removed**
No assets were removed

**Bigger**
No assets were bigger

**Smaller**
No assets were smaller

**Unchanged**
Asset | File Size | % Changed
----- | --------- | ---------
kcab.worker.CqRg4sOw.js | 4.65 MB | 0%
</div>
</details>

---

#### api

**Total**
Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
2 | 3.87 MB | 0%

<details>
<summary>View detailed bundle breakdown</summary>
<div>

**Added**
No assets were added

**Removed**
No assets were removed

**Bigger**
No assets were bigger

**Smaller**
No assets were smaller

**Unchanged**
Asset | File Size | % Changed
----- | --------- | ---------
index.js | 3.87 MB | 0%
models.js | 0 B | 0%
</div>
</details>

---

#### cli

**Total**
Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
1 | 8.02 MB | 0%

<details>
<summary>View detailed bundle breakdown</summary>
<div>

**Added**
No assets were added

**Removed**
No assets were removed

**Bigger**
No assets were bigger

**Smaller**
No assets were smaller

**Unchanged**
Asset | File Size | % Changed
----- | --------- | ---------
cli.js | 8.02 MB | 0%
</div>
</details>

---

#### crdt

**Total**
Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
1 | 11.16 kB | 0%

<details>
<summary>View detailed bundle breakdown</summary>
<div>

**Added**
No assets were added

**Removed**
No assets were removed

**Bigger**
No assets were bigger

**Smaller**
No assets were smaller

**Unchanged**
Asset | File Size | % Changed
----- | --------- | ---------
index.js | 11.16 kB | 0%
</div>
</details>
</details>

### CODE REVIEW

**Comment by nikhilweee** [comment]
Hi folks, anything I can do to move this forward?

**Comment by nikhilweee** [comment]
Hi everyone, bumping this again. Would love to know how to move this forward.

**Comment by matt-fidd** [comment]
Hi @nikhilweee 

It looks like the tests are failing. The update-vrt command isn't working because it isn't the visuals failing, take a look at the log in https://github.com/actualbudget/actual/actions/runs/26696755365/job/78682581233?pr=7749 for more info.

I'm also not seeing the same behaviour as you on the most recent version of Actual. Pressing escape does revert the changes in all of the tables I've tried it in, which ones are you seeing an issue with? I agree it doesn't drop you out of edit mode.

**Comment by nikhilweee** [comment]
@matt-fidd Thanks for reviewing! I fixed a tricky event timing issue where `TagAutocomplete` was swallowing the `Escape` key, causing the underlying `<Input>` to mistakenly save the dirty text on blur. I stopped `<Input>` from saving natively on blur and ensured `Escape` is properly forwarded to `NotesCell` so it can discard the temporary text. This should be ready for review!

**Comment by nikhilweee** [comment]
Hey @matt-fidd this should be ready for another review. I see the deploy preview pipeline has been failing for other PRs too, so that looks like a separate issue.

**Comment by nikhilweee** [comment]
Hey @matt-fidd, gentle nudge to take a look at this one again!

File: packages/desktop-client/src/components/transactions/TransactionsTable.tsx [hunk 1]
```diff
@@ -1978,6 +1980,12 @@ function NotesCell({
     if (e.key === 'Enter' || e.key === 'Tab') {
       onUpdate(inputValue);
     } else if (e.key === 'Escape') {
+      if (!e.isPropagationStopped()) {
+        escapePressed.current = true;
```
Comment (matt-fidd) [inline packages/desktop-client/src/components/transactions/TransactionsTable.tsx:1984]: this never gets reset does it? It feels wrong for it to stay true, that means the guard onBlur only runs once

File: packages/desktop-client/src/components/autocomplete/TagAutocomplete.tsx [hunk 2]
```diff
@@ -141,7 +140,13 @@ export function TagAutocomplete({
       e.stopPropagation();
       void handleSelect(highlightedId);
     } else if (e.key === 'Escape') {
+      if (showPopup) {
```
Comment (matt-fidd) [inline packages/desktop-client/src/components/autocomplete/TagAutocomplete.tsx:143]: This is always true at this point, it's guarded https://github.com/actualbudget/actual/pull/7749/changes#diff-5e695312cc11c25c92b54955e6eba6fb2a4787611a75ed342a0fe85ee971da4cR119-R121

File: packages/desktop-client/src/components/table.tsx [hunk 3]
```diff
@@ -1332,6 +1355,24 @@ export function useTableNavigator<T extends TableItem>(
...
+        );
+        if (idx !== -1) {
+          nextIdx = idx;
+        }
+      }
+
+      const next = data[nextIdx];
+      const availableFields = getFields(next);
+      if (availableFields.length > 0) {
+        onEdit(next.id, availableFields[0]);
```
Comment (matt-fidd) [inline packages/desktop-client/src/components/table.tsx:1374]: we don't always want to resume editing on the first available field do we? I think if we're going to implement auto-resume it should pick up from where the user left off

File: packages/desktop-client/src/components/transactions/TransactionsTable.tsx [hunk 4]
```diff
@@ -1978,6 +1980,12 @@ function NotesCell({
...
+      }
+      if (e.target && 'value' in e.target) {
+        (e.target as HTMLInputElement).value = note;
+      }
```
Comment (matt-fidd) [inline packages/desktop-client/src/components/transactions/TransactionsTable.tsx:1988]:
```suggestion
      if (e.target instanceof HTMLInputElement) {
        e.target.value = note;
      }
```

**Review by matt-fidd — CHANGES_REQUESTED** [review]
Some things I noticed while testing:
- Pressing escape on an edited date cell will not revert the date, it commits the new one
- Pressing escape in an amount field of the "Add New" editor deletes the entire transactionn 
- You need to click escape twice after clicking the cleared checkbox

**Comment by matt-fidd** [comment]
Apologies that it took a while to get a review on this, it's been hard to work out how to test it properly

**Comment by nikhilweee** [comment]
@matt-fidd thanks for your review. I'll take a look as soon as I have some time.

File: packages/desktop-client/src/components/table.tsx [hunk 5]
```diff
@@ -1244,6 +1264,20 @@ export type TableNavigator<T extends TableItem> = {
...
+// keyboard navigation working, but nothing is actually being edited.
+const NON_EDITABLE_FIELDS = new Set([
+  'select',
+  'cleared',
+  'cancel',
+  'add',
+  'schedule',
+]);
+
+export function isEditingCell(editingId: unknown, focusedField: string) {
```
Comment (MikesGlitch) [inline packages/desktop-client/src/components/table.tsx]: editingId has this type I think

`TableItem['id'] | null`

**Comment by nikhilweee** [comment]
Hi @matt-fidd @MikesGlitch,

I have merged the latest `master` and refactored the table editing and navigation logic to make the Escape/edit lifecycle robust:

- **Navigator state separation**: `useTableNavigator` now cleanly distinguishes cursor focus from active editing (`isEditing`, `cursor`, `beginEdit`, `exitEdit`), making navigation vs. editing state explicit.
- **Unified cell edit lifecycle**: Introduced `useCellEditor` and `editSession` for table cells (`InputValue`, `CustomCell`, `DateSelect`), giving cells explicit session states (`active`, `committed`, `canceled`). This replaces previous ad-hoc event flags and ensures Escape cleanly reverts drafts without race conditions on blur.
- **Edge cases resolved**: Addressed Escape handling in `DateSelect` and duplicate rename dispatches in tag rows.
- **Full test coverage & green CI**: All 39 CI checks—including full monorepo unit tests and the Playwright `keyboard-navigation` E2E suite—are passing.

Ready for another review whenever you have time!
Fixes https://github.com/actualbudget/actual/issues/3362
